### PR TITLE
refactor: Separate generation of .rst and .py files for Datamodel and TUI

### PR DESCRIPTION
**Pull request by hpohekar**
Closes https://github.com/ansys/pyfluent/issues/2661 https://github.com/ansys/pyfluent/issues/1192

1. `doc/datamodel_rstgen.py` generates datamodel .rst files in `doc/api/meshing/datamodel` and `doc/api/solver/datamodel`
2. `doc/tui_rstgen.py` generates TUI .rst files in `doc/api/meshing/tui` and `doc/api/solver/tui`
3. Added `datamodel_rstgen.py` and `tui_rstgen.py` into `Makefile`.

### CODE REVIEW

File: doc/Makefile [hunk 1]
```diff
@@ -23,4 +23,8 @@ clean:
 	rm -rf build_errors.txt
 	rm -rf $(BUILDDIR)/*
 	rm -rf $(SOURCEDIR)/examples
-	find . -type d -name "_autosummary" -exec rm -rf {} +
+	rm -rf $(SOURCEDIR)/api/meshing/datamodel
+	rm -rf $(SOURCEDIR)/api/meshing/tui
+	rm -rf $(SOURCEDIR)/api/solver/datamodel
+	rm -rf $(SOURCEDIR)/api/solver/tui
```
Comment (hpohekar) [inline doc/Makefile:29]: `.\make clean` was not removing `datamodel/*` and `tui/*` .rst files. Now we can remove them using `.\make clean`.

File: doc/make.bat [hunk 2]
```diff
@@ -35,6 +35,10 @@ goto end
 :clean
 rmdir /s /q %BUILDDIR% > /NUL 2>&1
 rmdir /s /q %SOURCEDIR%\examples > /NUL 2>&1
+rmdir /s /q %SOURCEDIR%\api\meshing\datamodel > /NUL 2>&1
+rmdir /s /q %SOURCEDIR%\api\meshing\tui > /NUL 2>&1
+rmdir /s /q %SOURCEDIR%\api\solver\datamodel > /NUL 2>&1
+rmdir /s /q %SOURCEDIR%\api\solver\tui > /NUL 2>&1
```
Comment (hpohekar) [inline doc/make.bat:41]: `.\make clean` was not removing `datamodel/*` and `tui/*` .rst files. Now we can remove them using `.\make clean`.

File: Makefile [hunk 3]
```diff
@@ -126,6 +126,8 @@ build-doc-source:
 	@xvfb-run make -C doc html
 
 build-all-docs:
+	@python doc/datamodel_rstgen.py
+	@python doc/tui_rstgen.py
```
Comment (hpohekar) [inline Makefile:130]: Added here for doc build.

File: doc/datamodel_rstgen.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+"""Provides a module to generate the Fluent Datamodel rst files."""
```
Comment (PipKat) [inline doc/datamodel_rstgen.py:1]:
```suggestion
"""Provides a module for generating the Fluent datamodel result files."""
```
Use plain English for rst files.

File: doc/tui_rstgen.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+"""Provides a module to generate the Fluent TUI rst files."""
```
Comment (PipKat) [inline doc/tui_rstgen.py:1]:
```suggestion
"""Provides a module for generating Fluent TUI result files."""
```

File: codegen/datamodelgen.py [hunk 6]
```diff
@@ -463,11 +481,19 @@ def _delete_generated_files(self):
             shutil.rmtree(Path(_SOLVER_DM_DOC_DIR))
 
 
-def generate(version, pyfluent_path, sessions: dict):
-    return DataModelGenerator(version, pyfluent_path, sessions).write_static_info()
+def generate(
```
Comment (PipKat) [inline codegen/datamodelgen.py:484]: Don't forget to add "Parameters" and "Returns" sections with the appropriate descriptions.

Comment (hpohekar) [reply codegen/datamodelgen.py:438]: Done. Thank you.

File: codegen/allapigen.py [hunk 7]
```diff
@@ -58,7 +58,7 @@ def _update_first_level(d, u):
     )
     _update_first_level(
         api_tree,
-        datamodelgen.generate(version, args.pyfluent_path, sessions),
+        datamodelgen.generate(version, args.pyfluent_path, sessions, False, True),
```
Comment (seanpearsonuk) [inline codegen/allapigen.py:61]:
```suggestion
        datamodelgen.generate(version, args.pyfluent_path, sessions, generate_rst=False, generate_py=True),
```

File: codegen/datamodelgen.py [hunk 8]
```diff
@@ -162,7 +166,8 @@ def __init__(
             )
         if not self._static_info["solverworkflow"]:
             del self._static_info["solverworkflow"]
-        self._delete_generated_files()
+        if self.generate_py:
+            self._delete_generated_files()
```
Comment (seanpearsonuk) [inline codegen/datamodelgen.py:170]: Please can we add a comment to clarify this logic. It can help future maintenance.

Comment (hpohekar) [reply codegen/datamodelgen.py:170]: Yes, will do.

File: codegen/datamodelgen.py [hunk 9]
```diff
@@ -463,11 +481,39 @@ def _delete_generated_files(self):
...
+    Parameters
+    ----------
+    version: str
+        Fluent version.
+    pyfluent_path: str
+        File generation path.
+    sessions: dict
+        Fluent sessions.
+    generate_rst: bool
+        If ``True`` then generates .rst files.
```
Comment (PipKat) [inline codegen/datamodelgen.py:502]:
```suggestion
        Whether to generate result (RST) files.
```

File: codegen/datamodelgen.py [hunk 10]
```diff
@@ -463,11 +481,39 @@ def _delete_generated_files(self):
...
+    generate_py: bool
+        If ``True`` then generates .py files.
```
Comment (PipKat) [inline codegen/datamodelgen.py:504]:
```suggestion
        Whether to generate Python (PY) files.
```

File: doc/datamodel_rstgen.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+"""Provides a module for generating the Fluent datamodel result files."""
+
+from ansys.fluent.core import FluentMode, launch_fluent
+from ansys.fluent.core.utils.fluent_version import get_version_for_file_name
+from codegen.datamodelgen import generate
+
+if __name__ == "__main__":
+    sessions = {FluentMode.SOLVER: launch_fluent()}
+    version = get_version_for_file_name(session=sessions[FluentMode.SOLVER])
+    generate(version, None, sessions, True, False)
```
Comment (mkundu1) [inline doc/datamodel_rstgen.py:10]: Will this launch Fluent? We have to do the docgen without launching Fluent. I think the generated API classes have all information to write out the .rst files.

File: doc/tui_rstgen.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    )
+
+
+def _get_tui_file_name(mode: str):
+    for file in os.listdir(_get_tui_filepath(mode)):
+        if fnmatch.fnmatch(file, "tui_*.py"):
+            return file
+
+
+def generate():
```
Comment (mkundu1) [inline doc/tui_rstgen.py:45]: Does this work even if we don't generate the hierarchy of .rst files/directories?

File: doc/datamodel_rstgen.py (hunk 11 above)
Comment (hpohekar) [reply doc/datamodel_rstgen.py:10]: We are going to remove this code. We will discuss this soon.

File: doc/tui_rstgen.py (hunk 12 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:45]: There is only single file inside each of `doc/source/api/meshing/tui/index.rst` and `doc/source/api/solver/tui/index.rst` therefore we are creating that single `index.rst` file using Python.  

We don't have hierarchy of .rst files/directories for tui, it is for datamodel only.

File: codegen/datamodelgen.py (hunk 8 above)
Comment (hpohekar) [reply codegen/datamodelgen.py:170]: We are going to remove this code.

File: doc/tui_rstgen.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    """Generates TUI .rst files."""
```
Comment (PipKat) [inline doc/tui_rstgen.py:46]:
```suggestion
    """Generate TUI result (RST) files."""
```

File: codegen/datamodelgen.py [hunk 14]
```diff
@@ -463,11 +481,39 @@ def _delete_generated_files(self):
...
-def generate(version, pyfluent_path, sessions: dict):
-    return DataModelGenerator(version, pyfluent_path, sessions).write_static_info()
+def generate(
+    version: str,
+    pyfluent_path: str,
+    sessions: dict,
+    generate_rst: bool,
+    generate_py: bool,
+):
+    """Generates datamodel codegen files.
```
[the remark this answers is not in the document]
Comment (hpohekar) [reply codegen/datamodelgen.py:491]: "codegen" is used for "Code Generation" which generates .PY files for datamodel, TUI and settings.

[the remark this answers is not in the document]
Comment (hpohekar) [reply codegen/datamodelgen.py:491]: We will change it to `Generate datamodel files`. Thank you.

[the remark this answers is not in the document]
Comment (hpohekar) [reply codegen/datamodelgen.py]: Done. Thank you.

**Review by seanpearsonuk — APPROVED** [review]

**Comment by seanpearsonuk** [comment]
@hpohekar Ready to merge?  Should this be marked _Ready for review_?

**Comment by hpohekar** [comment]
> @hpohekar Ready to merge? Should this be marked _Ready for review_?

No. It's WIP.

File: codegen/datamodelgen.py [hunk 15]
```diff
@@ -463,11 +481,39 @@ def _delete_generated_files(self):
...
+    generate_py: bool,
+):
+    """Generate datamodel files.
+
+    Parameters
+    ----------
+    version: str
+        Fluent version.
+    pyfluent_path: str
+        File generation path.
```
Comment (PipKat) [inline codegen/datamodelgen.py:498]:
```suggestion
        Path to save generated datamodel files to.
```

File: doc/datamodel_rstgen.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+"""Provides a module for generating the Fluent datamodel result files."""
```
Comment (PipKat) [inline doc/datamodel_rstgen.py:1]:
```suggestion
"""Provides a module for generating Fluent datamodel result files."""
```

Comment (hpohekar) [reply doc/datamodel_rstgen.py:1]: Done. Thank you.

File: doc/tui_rstgen.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
+"""Provides a module for generating Fluent TUI .rst files."""
```
Comment (PipKat) [inline doc/tui_rstgen.py:1]:
```suggestion
"""Provides a module for generating Fluent TUI result files."""
```
Use result like was done for in the previous file.

File: doc/tui_rstgen.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+    attribute_classes = []
+    attributes_dict = dict(vars(menu))
+    for attr_name, attr_value in attributes_dict.items():
+        if not attr_name.startswith("__"):
+            attribute_classes.append(attributes_dict[attr_name])
+    return attribute_classes
+
+
+def _get_attribute_classes_with_and_without_members(menu=None):
+    """Get classes with and without sub members as attributes.
```
Comment (PipKat) [inline doc/tui_rstgen.py:36]:
```suggestion
    """Get classes with and without submembers as attributes.
```

File: doc/tui_rstgen.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+    )
+
+
+def _get_tui_file_name(mode: str):
+    """Get tui_*.py file name.
+
+    Parameters
+    ----------
+    mode: str
+        Fluent session mode either ``meshing`` or ``solver``.
```
Comment (PipKat) [inline doc/tui_rstgen.py:181]:
```suggestion
        Fluent session mode. Options are ``meshing`` and ``solver``.
```

File: doc/tui_rstgen.py [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+def _write_doc(menu, mode):
+    """Write .rst file for each menu.
+
+    Parameters
+    ----------
+    menu: type
+        TUIMenu, TUIMethod
+    mode: str
+        Fluent session mode either ``meshing`` or ``solver``.
```
Comment (PipKat) [inline doc/tui_rstgen.py:211]:
```suggestion
        Fluent session mode. Options are ``meshing`` and ``solver``.
```

File: doc/tui_rstgen.py [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+        dict(
+            name=main_menu, with_members=with_members, without_members=without_members
+        ),
+    )
+    for member in with_members:
+        _get_attribute_classes_recursively(member, Menus)
+
+
+def _generate_doc(Menus, mode):
+    """Write .rst file for each attribute class.
```
Comment (PipKat) [inline doc/tui_rstgen.py:263]:
```suggestion
    """Write result file for each attribute class.
```

File: doc/tui_rstgen.py [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+
+    Parameters
+    ----------
+    Menus: type
+        ``Menus`` class.
+    mode: str
+        Fluent session mode either ``meshing`` or ``solver``.
```
Comment (PipKat) [inline doc/tui_rstgen.py:270]:
```suggestion
        Fluent session mode. Options are ``meshing`` and ``solver``.
```

File: doc/tui_rstgen.py [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+    """
+    all_menus_attrs = [attr for attr in dir(Menus) if not attr.startswith("__")]
+    for attr in all_menus_attrs:
+        menu_attr = getattr(Menus, attr)
+        _write_doc(menu_attr, mode)
+
+
+def generate(main_menu, mode):
+    """Generate .rst files.
```
Comment (PipKat) [inline doc/tui_rstgen.py:279]:
```suggestion
    """Generate result files.
```

File: doc/tui_rstgen.py [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+
+    Parameters
+    ----------
+    main_menu: type
+        ``main_menu`` class.
+    mode: str
+        Fluent session mode either ``meshing`` or ``solver``.
```
Comment (PipKat) [inline doc/tui_rstgen.py:286]:
```suggestion
        Fluent session mode. Options are ``meshing`` and ``solver``.
```

Comment (hpohekar) [reply doc/tui_rstgen.py:317]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 23 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:310]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 22 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:301]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 21 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:294]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 20 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:244]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 19 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:210]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 17 above)
Comment (hpohekar) [reply doc/tui_rstgen.py:1]: Done. Thank you.

File: doc/tui_rstgen.py (hunk 23 above)
Comment (mkundu1) [reply doc/tui_rstgen.py]: .rst are not results file. `.rst` is fine or we can say - `Generate RST files.`

Comment (hpohekar) [reply doc/tui_rstgen.py]: Done. Thank you.

**Review by mkundu1 — APPROVED** [review]

**Review by prmukherj — APPROVED** [review]

File: tests/test_flobject.py [hunk 25]
```diff
@@ -1070,6 +1070,19 @@ def test_ansys_units_integration_nested_state(load_mixing_elbow_mesh):
...
+            "reference_frame": "Absolute",
+            "velocity": {"option": "value", "value": (0, "m s^-1")},
+            "velocity_specification_method": "Magnitude, Normal to Boundary",
+        },
+        "name": "hot-inlet",
+        "turbulence": {
+            "turbulent_specification": "Intensity and Viscosity Ratio",
+            "turbulent_intensity": (0.05, ""),
+            "turbulent_viscosity_ratio": (10, None),
+        },
```
Comment (hpohekar) [inline tests/test_flobject.py:1085]: Order of keys is changing.

**Review by seanpearsonuk — APPROVED** [review]

**Pull request merged**